### PR TITLE
Expose prometheus with basic auth

### PR DESCRIPTION
**Pull request by yuvipanda**
This is the beginning of implementing idea 1 from the
list @georiganaelena made in
https://github.com/2i2c-org/infrastructure/issues/328#issuecomment-1016601966.

We have one prometheus running per cluster, but manage many clusters.
A single grafana that can connect to all such prometheus clusters
will help with monitoring as well as reporting. So we need to expose
it as securely as possible to the external world, as it can contain
private information.

In this case, we're using https + basic auth provided by
nginx-ingress
(https://kubernetes.github.io/ingress-nginx/examples/auth/basic/)
to safely expose prometheus to the outside world. We can then
use a grafana that knows these username / passwords to access this
prometheus instance. Each cluster needs its own username / password
(generated with pwgen 64 1), so users in one cluster can not access
prometheus for another cluster.

Ref https://github.com/2i2c-org/infrastructure/issues/328

### CODE REVIEW

**Comment by yuvipanda** [comment]
I used a tiny helper script to generate these encrypted secret files:

```
#!/bin/bash
set -euo pipefail

F="${1}/enc-support.secret.yaml"
echo "prometheusAuth:" > $F
echo "  username: $(pwgen 64 1)" >> $F
echo "  password: $(pwgen 64 1)" >> $F
sops -i -e $F
echo "Done $F"%   
```

Then ran `find . -type d | xargs -L1 ./gen.bash  ` to run it for all dirs inside config/clusters.

However, meom-ige and farallon still don't have a support cluster, so this can't be used there yet.

**Comment by yuvipanda** [comment]
Note - I've already run `deploy-support` for all these :)

**Comment by consideRatio** [comment]
Hmmm, the support chart could also do the trick jupyterhub does and generate secrets in a Secret file - if prometheus could accept accessing the content from there.

It may not be qhat makes sense, just floating the idea to ensure it is considered as an option. Someplace, the grafana server would need to have access to those secrets in other clusters so it would probably be relevant to have centrally still.

**Comment by yuvipanda** [comment]
@consideRatio yeah, that's the other option - but that would require giving kubernetes API access to every single cluster to the centralized grafana, which I would very much prefer to not do.

File: deployer/cluster.py
Comment (consideRatio) [inline deployer/cluster.py:11]:
```suggestion
from file_acquisition import get_decrypted_files
```

File: deployer/cluster.py
Comment (consideRatio) [inline deployer/cluster.py:76]: Reminder to add this back

File: deployer/cluster.py
Comment (consideRatio) [inline deployer/cluster.py:227]: A small bugfix, nice!

File: helm-charts/support/values.schema.yaml
Comment (consideRatio) [inline helm-charts/support/values.schema.yaml:73]:
```suggestion
  # prometheusAuth is _not a dependent helm chart_. It is values directly
  # coupled with this helm chart and are influencing the rendering of templates
  # we provide as part of this helm chart.
  #
  # Due to that, we maintain a strict schema here but not elsewhere.
  #
  prometheusAuth:
```

File: helm-charts/support/values.yaml
Comment (consideRatio) [inline helm-charts/support/values.yaml:38]: This setup enables a nice security feature I think.

By using a netpol for prometheus internally that requires traffic to arrive from the ingress controller pods which in turn enforce passwords, we would have locked down access to prometheus internally and externally.

File: helm-charts/support/values.yaml
Comment (consideRatio) [inline helm-charts/support/values.yaml:35]: TODO: This secret needs to be created and consume the values passed.

File: helm-charts/support/values.yaml
Comment (consideRatio) [inline helm-charts/support/values.yaml:1]: The naming `prometheusAuth` is helpful, but I don't think it is helpful enough. We who use this configuration this often need to know more technical details and understand what this influence. I think a more detailed description is favorable, hiding less of the technical details and by that also giving more insights into what is done by configuring this.

I got confused about `prometheusAuth` as if it was how you configured prometheus to require authentication, but in reality the strategy is coupled with Prometheus creating an Ingress resource to be managed by nginx-ingress which in turn is hardcoded to reference a k8s Secret named `prometheus-ingress-auth` for username/password credentials required for requests.

I think this name should reflect that this is related to a) prometheus authentication, b) it is a k8s Secret being created, c) it is referenced by an Ingress resource, and d) the ingress resource is managed by nginx-ingress.

`prometheusAuthIngressSecret` perhaps?

**Review by consideRatio — COMMENTED** [review]
This is a nice beginning!

### My understanding summarized

1. We assume we have `prometheus` subdomains for each cluster point to an `nginx-ingress` controllers in each cluster.
2. Letting the support chart create a k8s Secret to hold username/password for prometheus access via support chart values.
3. Letting the support chart's dependent prometheus Helm chart create an k8s Ingress resource that references the created k8s Secret with username/password. This reference is a annotation that an nginx-ingress controller understands: we wish to require basic authentication using the provided username/password in the k8s Secret
4. Later we manually (?) configure a central Grafana instance through its web UI to have many data sources, one for each cluster. Following this, we can allow dashboards there to choose to inspect any given cluster or have some dashboards that summarize the state of all clusters.

### A few action points

- [x] Add a k8s Secret template in the support chat
- [ ] Do work to ensure we get the relevant new sub-domain names get CNAME entries in the DNS servers we control, I figure they should point to the primary domain name we use which also points to the ingress controllers IP.
- [ ] Point out a specific Grafana instance to be the central instance we use (the one in the 2i2c cluster right?)

File: config/clusters/azure.carbonplan/support.values.yaml
Comment (consideRatio) [inline config/clusters/azure.carbonplan/support.values.yaml:4]: There is a CI error about this. There are two separate entries for `prometheus` in this file, where one below configures the resource requests.

File: deployer/cluster.py
Comment (yuvipanda) [reply deployer/cluster.py:11]: Done!

File: helm-charts/support/values.schema.yaml
Comment (yuvipanda) [reply helm-charts/support/values.schema.yaml]: Done!

File: helm-charts/support/values.yaml
Comment (yuvipanda) [reply helm-charts/support/values.yaml:35]: @consideRatio oh yeah, I just forgot to git add that file! done :)

File: helm-charts/support/values.yaml
Comment (yuvipanda) [reply helm-charts/support/values.yaml]: @consideRatio flawless logic and very well written feedback! I've renamed it to be prometheusIngressAuthSecret (since it's the prometheusIngress we are providing auth to, not prometheus itself).

File: config/clusters/azure.carbonplan/support.values.yaml
Comment (yuvipanda) [reply config/clusters/azure.carbonplan/support.values.yaml:7]: Thanks for catching it! Fixed! All hail CI!

File: deployer/cluster.py
Comment (yuvipanda) [reply deployer/cluster.py:227]: ty :)

File: helm-charts/support/templates/prometheus-htpasswd.yaml
Comment (consideRatio) [inline helm-charts/support/templates/prometheus-htpasswd.yaml:10]: - Remove blank new line in the beginning of file
- Add new line in end of file
- Add right whitespace chomp on the helm template expression before the first content has been rendered as a rule of thumb to ensure we don't render a blank line as the first thing as our output of the rendered helm template.

```suggestion
{{- if .Values.prometheusIngressAuthSecret.enabled -}}
apiVersion: v1
kind: Secret
metadata:
  name: prometheus-ingress-auth-basic
type: Opaque
data:
  auth: {{ htpasswd .Values.prometheusIngressAuthSecret.username .Values.prometheusIngressAuthSecret.password | b64enc | quote}}
{{- end }}

```

File: helm-charts/support/values.yaml
Comment (consideRatio) [inline helm-charts/support/values.yaml:37]:
```suggestion
        # under prometheusIngressAuthSecret. Ingress is not enabled by default,
```

File: helm-charts/support/values.yaml
Comment (consideRatio) [inline helm-charts/support/values.yaml:36]: I don't understand that "this secret is created via the prometheus-htpasswd.yaml file, can you elaborate?

To clarify my confusion, I'm thinking `"The referenced k8s Secret is created by the Helm chart via config under prometheusIngressBasicAuth"`, and I lack knowledge about a `prometheus-htpasswd.yaml` file. I figure the k8s Secret will be consumed by the ingress controller independent of any prometheus logic.

Comment (yuvipanda) [reply helm-charts/support/values.yaml:36]: @consideRatio https://github.com/2i2c-org/infrastructure/pull/1091/files#diff-7042b93ab64b7c665e0172a141323f9e75f60d0fc7d06fce61dad6bc4ba41fff is prometheus-htpasswd.yaml, the actual file that is creating the k8s secret.

Comment (yuvipanda) [reply helm-charts/support/values.yaml]: Clarified by renaming file as well as saying `templates/<filename>`

Comment (consideRatio) [reply helm-charts/support/values.yaml]: Oh when you said "the file that is creating the k8s secret" I got thinking something like `kubectl create secret generic --from-file ...`. But I understand it is a Helm chart template file that renders the k8s Secret resource! I figure if we clarify that it is the file mentioned is a helm template, it would be a lot clearer to me.

Can we also name the helm chart template file with a suffix of `secret.yaml`? It is very common practice in helm charts to name template files in a way that includes the k8s resource type, and it has been written in the Helm docs in the past at least that they suggest one resource per template file. Currently I see they have updated [their docs](https://helm.sh/docs/chart_template_guide/getting_started/) to say the less strict advice of:

> TIP: Template names do not follow a rigid naming pattern. However, we recommend using the suffix .yaml for YAML files and .tpl for helpers.

Comment (yuvipanda) [reply helm-charts/support/values.yaml]: @consideRatio ah, I'm not a big fan of adding the type of resource being created to the name of the file, preferring to signal intent instead. Just naming it the same as resource type works great when you're deploying a single application, or when it's under a subdirectory whose name explains what is being deployed. I'd like to not have both a name that explains intent and a suffix. It also reminds me a lot of [hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation). I'd prefer to keep it out if that's ok with you.

I've reworded the comment to make it clearer both where the secret is generated, as well as what feeds content in there.

File: helm-charts/support/templates/prometheus-ingress-auth.yaml
Comment (consideRatio) [inline helm-charts/support/templates/prometheus-ingress-auth.yaml:2]:
```suggestion
{{- if .Values.prometheusIngressAuthSecret.enabled -}}
```

File: helm-charts/support/templates/prometheus-ingress-auth.yaml
Comment (consideRatio) [inline helm-charts/support/templates/prometheus-ingress-auth.yaml:10]:
```suggestion
{{- end }}

```

**Review by consideRatio — COMMENTED** [review]
This LGTM, if needed, lets iterate from here to not slow down too much!

I think the key points now with relation to this PR is that this goes with the creation of additional domain names etc and those aspects may need documentation updates.

File: helm-charts/support/values.yaml
Comment (consideRatio) [reply helm-charts/support/values.yaml]: Oh okay - It gives me some itch as the convention is strong and has never been broken in the helm chart's I recall seeing besides the occational `rbac.yaml` file that includes multiple rbac related resource kinds - which in its name indicates what kind of resources it includes though.

Let's go with this though if you feel strongly about it!

Comment (yuvipanda) [reply helm-charts/support/values.yaml]: @consideRatio i've made a directory called prometheus-ingress-auth and moved this to a secret.yaml file there. I think this should both satisfy your itch as well as my desire to avoid hungarian notation.

Comment (consideRatio) [reply helm-charts/support/values.yaml]: :D You are a kind person Yuvi!

File: helm-charts/support/templates/prometheus-htpasswd.yaml
Comment (yuvipanda) [reply helm-charts/support/templates/prometheus-htpasswd.yaml]: Done!

File: helm-charts/support/templates/prometheus-ingress-auth.yaml
Comment (yuvipanda) [reply helm-charts/support/templates/prometheus-ingress-auth.yaml]: Done, I think!

File: helm-charts/support/templates/prometheus-ingress-auth.yaml
Comment (yuvipanda) [reply helm-charts/support/templates/prometheus-ingress-auth.yaml]: Done!

File: helm-charts/support/values.yaml
Comment (yuvipanda) [reply helm-charts/support/values.yaml]: @consideRatio 
image

**Comment by yuvipanda** [comment]
I've added docs on what needs to happen with DNS, and I think it's already happened for all the clusters mentioned here.

File: docs/howto/operate/grafana.md
Comment (consideRatio) [inline docs/howto/operate/grafana.md:35]:
```suggestion
prometheusIngressAuthSecret:
```

File: docs/howto/operate/grafana.md
Comment (consideRatio) [inline docs/howto/operate/grafana.md:63]: I think it will be critical that this is named `enc-support.secret.values.yaml` --- that it includes `values.yaml` in the end to be compatible with https://github.com/2i2c-org/infrastructure/pull/1093/files#diff-6d3cb43608c85f4e234a8decb5a6dc2f7f62e5606891b377941a201fd8222d25R77.

```suggestion
feature of nginx-ingress is used to restrict this. A `enc-support.secret.values.yaml`
```

File: docs/howto/operate/grafana.md
Comment (consideRatio) [inline docs/howto/operate/grafana.md:89]: TODO: Systematic rename of this file to `enc-support.secret.values.yaml`

File: helm-charts/support/values.yaml
Comment (consideRatio) [inline helm-charts/support/values.yaml:37]:
```suggestion
        # and the contents are controlled by config under prometheusIngressAuthSecret.
```

**Review by consideRatio — APPROVED** [review]
With these comments resolved, this LGTM! Nice work on this @yuvipanda!

File: docs/howto/operate/grafana.md
Comment (yuvipanda) [reply docs/howto/operate/grafana.md:63]: I thought the secret.yaml is an alternative to .values.yaml, specifying what they are. Why have both?

Comment (yuvipanda) [reply docs/howto/operate/grafana.md:63]: Either way, happy to just do this for now and move the conversation about conventions elsewhere :)

File: docs/howto/operate/grafana.md
Comment (yuvipanda) [reply docs/howto/operate/grafana.md]: Just done.

File: deployer/cluster.py
Comment (yuvipanda) [reply deployer/cluster.py]: Done!

File: deployer/cluster.py
Comment (GeorgianaElena) [inline deployer/cluster.py:115]: I believe this also needs to be restored?

File: deployer/cluster.py
Comment (consideRatio) [inline deployer/cluster.py:107]: I'll open a separate PR with this change after this is merged to avoid merge conflicts.

This is an optimization to not need to: a) add a named helm chart repo, b) updating all helm chart repos, c) referencing a helm chart repo via named helm chart repo. Instead one would simply name the helm chart and declare the repo, which is fetched just in time of the upgrade.

```suggestion
        print_colour("Provisioning cert-manager...")
        subprocess.check_call(
            [
                "helm",
                "upgrade",
                "cert-manager",  # given release name (aka. installation name)
                "cert-manager",  # helm chart to install
                f"--repo={cert_manager_url}",
                "--install",
                "--create-namespace",
                "--namespace=cert-manager",
                f"--version={cert_manager_version}",
                "--set=installCRDs=true",
```

File: deployer/cluster.py
Comment (consideRatio) [reply deployer/cluster.py:115]: Nice catch!!

Comment (consideRatio) [reply deployer/cluster.py:115]:
```suggestion
        subprocess.check_call(["helm", "dep", "up", support_dir])
```

File: deployer/cluster.py
Comment (yuvipanda) [reply deployer/cluster.py:107]: Perfect! Would definitely <3 a new PR!

**Comment by yuvipanda** [comment]
I think what's left here is to automate adding all these as a data source to one centralized grafana.

File: deployer/cluster.py
Comment (yuvipanda) [reply deployer/cluster.py]: aha! restored, thank you :)

**Comment by yuvipanda** [comment]
I think this is ready to go!

File: docs/howto/operate/grafana.md
Comment (sgibson91) [reply docs/howto/operate/grafana.md]: All values files can be secret, but not all secrets need to be values files. So I think we should adopt a `.secret.values.yaml` convention to indicate that this is a secret values file. See the notes box in the docs here: https://infrastructure.2i2c.org/en/latest/topic/config.html#id1

File: helm-charts/support/values.yaml
Comment (consideRatio) [reply helm-charts/support/values.yaml]: I opened https://github.com/2i2c-org/infrastructure/issues/1101 about this.

File: deployer/cluster.py
Comment (consideRatio) [reply deployer/cluster.py]: Retained memory of doing this in #1100

**Comment by yuvipanda** [comment]
Still need to write code that'll populate central grafana with datasources, but let's get this merged until then.

**Comment by damianavila** [comment]
> Still need to write code that'll populate central grafana with datasources

I think the upcoming work is actually described here: https://github.com/2i2c-org/infrastructure/issues/328#issuecomment-1082542191, let me know if you disagree.

Btw, thanks for all this work, @yuvipanda!!

**Pull request merged**
